File: src/main/java/io/github/mg138/tsbook/config/gui/ArmorGUIConfig.java
package io.github.mg138.tsbook.config.gui;

import io.github.mg138.tsbook.utils.Translate;
import io.github.mg138.tsbook.config.gui.element.ArmorGUIElementSetting;
import io.github.mg138.tsbook.config.gui.element.ArmorElementSetting;
import io.github.mg138.tsbook.config.gui.element.GUIElementSetting;
import org.bukkit.Material;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;

import java.util.*;

public class ArmorGUIConfig {
    private static final ArmorGUIConfig instance = new ArmorGUIConfig();

    public static ArmorGUIConfig getInstance() {
        return instance;
    }

    private ArmorGUIConfig() {
    }

    public final Map<Integer, GUIElementSetting> elementSettings = new HashMap<>();

    public void load(YamlConfiguration yaml, Translate translate) {
        elementSettings.clear();

        Set<String> keys = yaml.getKeys(false);
        for (String key : keys) {
            int i = Integer.parseInt(key);

            ConfigurationSection section = yaml.getConfigurationSection(key);
            if (section == null)
                throw new NullPointerException("Something wrong happened when iterating through the config!");

            Material material = Material.valueOf(section.getString("MATERIAL"));

            int count = section.getInt("COUNT");

            String name = translate.translate("NAME", null, section);

            List<String> lore = translate.translateList("LORE", null, section);
            lore = lore == null ? Collections.emptyList() : lore;

            int model = section.getInt("MODEL");

            GUIElementSetting elementSetting;
            if (section.contains("EQUIPMENT")) {
                ConfigurationSection equipment = section.getConfigurationSection("EQUIPMENT");
                if (equipment == null)
                    throw new NullPointerException("Something wrong happened when iterating through the config!");

                elementSetting = new ArmorGUIElementSetting(i, material, count, name, lore, model, new ArmorElementSetting(equipment));
            } else {
                elementSetting = new GUIElementSetting(i, material, count, name, lore, model);
            }

            elementSettings.put(i, elementSetting);
        }
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/config/gui/armor/ArmorGUIConfig.kt
package io.github.mg138.tsbook.config.gui.armor

import io.github.mg138.tsbook.item.attribute.ItemType
import io.github.mg138.tsbook.player.ArcticGlobalDataService
import io.github.mg138.tsbook.player.data.PlayerData
import io.github.mg138.tsbook.config.gui.armor.element.ArmorElementSetting
import io.github.mg138.tsbook.config.gui.armor.element.ArmorSetting
import io.github.mg138.tsbook.config.gui.element.GUIElementSetting
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.util.*

// todo wtf is this

/*
object ArmorGUIConfig : Iterable<MutableMap.MutableEntry<Int, GUIElementSetting>> {
    private val elementSettings: MutableMap<Int, GUIElementSetting> = HashMap()

    operator fun get(key: Int) = elementSettings[key]
    override fun iterator(): Iterator<MutableMap.MutableEntry<Int, GUIElementSetting>> = elementSettings.iterator()

    fun getByType(itemType: ItemType, player: Player): Int {
        val data = ArcticGlobalDataService.inst.getData<PlayerData>(player) ?: return -1

        elementSettings.forEach { (i, setting) ->
            if (!data.equipment.containsKey(i)) {
                if (setting is ArmorElementSetting) {
                    if (setting.setting.itemType == itemType) return i
                }
            }
        }

        return -1
    }

    fun load(yaml: YamlConfiguration) {
        elementSettings.clear()
        for (key in yaml.getKeys(false)) {
            val i = key.toInt()
            val section = yaml.getConfigurationSection(key)!!

            elementSettings[i] = when {
                section.contains("EQUIPMENT") -> ArmorElementSetting(
                    i, section, ArmorSetting(section.getConfigurationSection("EQUIPMENT")!!)
                )
                else -> GUIElementSetting(i, section)
            }
        }
    }

    fun unload() {
        elementSettings.clear()
    }
}

 */